test/acceptance/user.controller.acceptance.ts

```diff
@@ -130,6 +133,74 @@ describe('UserController', () => {
     });
   });
 
+  describe('authentication functions', () => {
+    it('login returns a valid token', async () => {
+      const newUser = await userRepo.create(user);
+      await client
+        .post('/users/login')
+        .send({email: newUser.email, password: newUser.password})
+        .expect(200)
+        .then(getToken);
+
+      function getToken(res: Response) {
+        // is it good to return the token as text?
+        const token = res.text;
+        expect(token).to.not.be.empty();
+      }
+    });
+
+    it('login returns an error when invalid credentials are used', async () => {
+      const newUser = await userRepo.create(user);
+      newUser.password = 'wrong password';
+      await client
+        .post('/users/login')
+        .send({email: newUser.email, password: newUser.password})
+        .expect(401);
+    });
+
+    it('/me returns the current user', async () => {
+      const newUser = await userRepo.create(user);
+      let token = '';
+      await client
+        .post('/users/login')
```

Comment:
Please create the token via UserRepository API, no need to make a (slow) HTTP request here.

Reply:
@bajtos The tests are in the acceptance test file, that's why they make HTTP calls, if you look at the other test cases in this file they do the same :)

Reply:
IMO, the tests should make HTTP calls only for things they are testing. For example, a test for `login` should call login via HTTP. Other tests should use a token created directly via JS/TS API, because token creation is not really a part of the tested scenario.

There are two reasons why I am advocating for this approach:
- Calling Repository API is simpler, it makes the code easier to read, and also tends to produce more helpful error message if something goes wrong.
- Calling Repository API directly makes the tests run faster. HTTP requests add a small overhead that adds up when there are many test cases to run.

Reply:
neat 👍 rewrote the test to leverage repository APIs

Reply:
> rewrote the test to leverage repository APIs

@jannyHou I don't see those changes in this commit. AFAICT, login is still invoked via `await client.post('/users/login'). Am I reading the diff incorrectly?

Reply:
@bajtos latency 🙈 we just pushed the new commit, maybe refresh again?

Reply:
I still see that test are performing login via REST API, not by calling Repository API.

@jannyHou PTAL again